src/types/Rule.ts

```diff
@@ -1,10 +1,32 @@
 import { Diagnostic } from "./Diagnostic";
 import { Lintable } from "./Lintable";
-import { RuleId, AsyncRuleId } from "./RuleConfig";
+import { LintState, toLintState } from "./LintState";
+import { RuleId } from "./RuleConfig";
 
-export interface Rule<R extends RuleId> {
+export interface Rule<
+  R extends RuleId,
+  D extends Diagnostic<R> = Diagnostic<R>
+> {
```

Comment:
The generic `D` isn't needed anymore, since we know that it's always `Diagnostic<R>`

```suggestion
export interface Rule<R extends RuleId> {
```